src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py

```diff
@@ -203,170 +187,97 @@ def apply(self, node, options=None):
-            loop_iterators.append(loop_iterator)
-            if array_reduction and idx != int(dimension_literal.value)-1:
-                array_iterators.append(loop_iterator)
-
-        # Initialise the temporary variable.
-        rhs = self._init_var(var_symbol)
-        lhs = reference.copy()
-        new_assignment = Assignment.create(lhs, rhs)
-        assignment.parent.children.insert(assignment.position, new_assignment)
-
-        array_indices = []
-        for idx in range(ndims):
-            array_indices.append(Reference(loop_iterators[idx]))
-        array_ref = ArrayReference.create(array_ref.symbol, array_indices)
-
-        statement = self._loop_body(
-            array_reduction, array_iterators, var_symbol, array_ref)
-
+            mask_ref = mask_ref_parent.children[mask_ref_index]
+
+        orig_lhs = node.ancestor(Assignment).lhs.copy()
+        array_refs = rhs.walk(ArrayReference)
+        lhs = array_refs[0].copy()
+        assignment = Assignment.create(lhs, rhs.detach())
+        # Replace existing code so the new code gets access to symbol
+        # tables etc.
+        orig_assignment = node.ancestor(Assignment)
+        orig_assignment.replace_with(assignment)
+
+        # Step 2 call nemoarrayrange2loop_trans to create loop bounds
+        # and array indexing (keeping track of where the new loop nest
+        # is created). Also deal with the mask if it exists.
         if mask_ref:
-            # A mask argument has been provided
-            for ref in mask_ref.walk(Reference):
-                # pylint: disable=unidiomatic-typecheck
-                if ref.name == array_ref.name and type(ref) is Reference:
-                    # The array is not indexed so it needs indexing
-                    # for the loop nest.
-                    shape = [Reference(obj) for obj in loop_iterators]
-                    reference = ArrayReference.create(ref.symbol, shape)
-                    ref.replace_with(reference)
-            statement = IfBlock.create(mask_ref.detach(), [statement])
-
-        for idx in range(ndims):
-            statement = Loop.create(
-                loop_iterators[idx].copy(), loop_bounds[idx][0].copy(),
-                loop_bounds[idx][1].copy(), Literal("1", INTEGER_TYPE),
-                [statement])
-
-        assignment.parent.children.insert(assignment.position, statement)
+            # add mask to the rhs of the assignment
+            assignment_rhs = BinaryOperation.create(
+                BinaryOperation.Operator.AND, assignment.rhs.copy(),
+                mask_ref.copy())
+            assignment.rhs.replace_with(assignment_rhs)
+
+        assignment_parent = assignment.parent
+        assignment_position = assignment.position
+        # Must be placed here to avoid circular imports
+        # pylint: disable=import-outside-toplevel
+        from psyclone.domain.nemo.transformations import \
+            NemoAllArrayRange2LoopTrans
```

Comment:
Again, try with the whole from path, I don't see why this imports should cycle (other than the folder has an aggregation of thing that might)

Reply:
There is a circular import problem here so it needs to remain as it is. The circular import makes sense I think as we are using a domain-specific transformation from a more generic PSyIR transformation and in general we would expect domain-specific transformations to import PSyIR transformations.